### ISSUE TITLE
Conformance Check: Promise Rejections Must be Caught 

### ISSUE THREAD

**Issue by ChadKillingsworth** [google/closure-compiler#1515]
I'd like to enforce that all promises have rejections caught. Either:

```js
promiseReturningFunction()
    .then(resFn, rejFn)
    .then(resFn) //warning here
    .then(resFn, rejFn);
```

or

```js
promiseReturningFunction()
    .then(resFn, rejFn)
    .then(resFn) //no warning here - it's caught
    .then(resFn, rejFn)
    .catch(rejFn);
```

I could also see this as a lint check rather than a conformance requirement.

**Comment by Dominator008** [comment]
This might be harder than it seems. At least it needs to be type aware if we want to avoid false positives.
See:
https://github.com/eslint/eslint/issues/1092
https://github.com/xjamundx/eslint-plugin-promise/blob/master/rules/always-catch.js

**Comment by ChadKillingsworth** [comment]
Since promises swallow exceptions without this, it is pretty critical.

**Comment by concavelenz** [comment]
A few comments:
* goog.Promise, throw uncaught exceptions (basically using a setTimeout to keep it out of the current chain of events). 

* That the first form you suggest as valid doesn't catch catch any exception in the last "resFn".  You almost always want an ".catch".

The real problem is that it is difficult to know when you are "done" with an promise.  You don't want to say this is bad:

/** @return {Promise<string>} */ function getSomething() {
   return doSomething().then(resFn);
}

or this:

var promise = doSomething();  // no catch

var x = promise.then(resFn);  // no catch
store(x);

var y = fetch(x);
y.catch(value);

 We could certainly say an "unconsumed" promise must have a "catch".  But even that would catch cases like this:


var initP;
function init() {
   // ... do something to 
   if (initP != null) {
     initP = new Promise(...);  // resolve when initialized
   }
   return initP;
}

init();  // start the init but don't care when it finishes.

So I'm not really sure what heuristic you are suggesting?

**Comment by ChadKillingsworth** [comment]
The vast majority of use cases is what I'm after - not every edge case. The eslint rule that @Dominator008 linked is what I'm most familiar with.

Everywhere a `.then` chain is used I would enforce this. Ideally it would be suppressible, but I realize conformance rules aren't suppressible as well.

**Comment by ChadKillingsworth** [comment]
Although I say that, but here's a case I'd like to handle as well:

```js
var a = Promise();
var b = Promise();

var c = Promise.all([a, b]);
```

**Comment by concavelenz** [comment]
I think any thenable should be checked, so referenced then check is reasonable.  As a lint check this is fine but the real solution is in the platform.  Top browsers are/have implemented the unhandled rejection handler.

**Comment by ChadKillingsworth** [comment]
> Top browsers are/have implemented the unhandled rejection handler.

Really? I haven't seen any chatter about that. That would be very awesome. Is there a spec/doc reference for that?

**Comment by concavelenz** [comment]
https://googlechrome.github.io/samples/promise-rejection-events/
